yesod-form/Yesod/Form/Fields.hs

```diff
@@ -637,6 +668,28 @@ optionsPairs opts = do
                  }
   return $ mkOptionList (zipWith mkOption [1 :: Int ..] opts)
 
+-- | Creates an 'OptionList' from a list of (display-value, internal value) pairs.
+optionsPairsGrouped
+  :: (MonadHandler m, RenderMessage (HandlerSite m) msg)
+  => [(msg, [(msg, a)])] -> m (OptionList a)
+optionsPairsGrouped opts = do
+  mr <- getMessageRender
+  let mkOption (external, (display, internal)) =
+          Option { optionDisplay       = mr display
+                 , optionInternalValue = internal
+                 , optionExternalValue = pack $ show external
+                 }
+      opts' = enumerateSublists opts -- :: [(msg, [(Int, (msg, a))])]
```

Comment:
Keep this comment? It helps to understand whats going on (as long it's not outdated). Much better would be a type annotation but `msg` and `a` seem to be not available here.

Same thing below at `yss`.

Reply:
If you use `ScopedTypeVariables` and an explicit `forall`, you can add in a type signature I believe. That would be preferable.